### PR TITLE
Mask2former hungarian matcher Add normalization before matmul

### LINKED ISSUE #25938: Mask2Former HungarianMatcherMaks2Former overflow on bad initial state when using 16-bit autocast
### System Info

- `transformers` version: 4.29.0
- Platform: Linux-5.15.0-79-generic-x86_64-with-glibc2.31
- Python version: 3.10.12
- Huggingface_hub version: 0.13.4
- Safetensors version: 0.3.2
- PyTorch version (GPU?): 2.0.0+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: yes

### Who can help?

@amy

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

1. import pair_wise_sigmoid_cross_entropy_loss from transformers.models.mask2former_modeling_mask2former
```
from transformers.models.mask2former.modeling_mask2former import pair_wise_sigmoid_cross_entropy_loss
```
2. initialize arbitrary input and labels. 
```
inputs = torch.randn([16,121445]).cuda() - 2 # To simulate bad initialization
labels = torch.randn([2, 121445]).cuda()
```
3. Compute loss with 16-bit autocast
```
with torch.autocast("cuda", dtype=torch.hafl):
     result = pair_wise_sigmoid_cross_entropy_loss(inputs, labels)
```
4. result got int everywhere
```
>>> result.isinf().all()
tensor(True, device='cuda:0')
```

I suggest normalizing `cross_entropy_loss_pos`, `cross_entropy_loss_neg` with `height_and_width` before computing matmul as follows:
```

def pair_wise_sigmoid_cross_entropy_loss(
    inputs: torch.Tensor, labels: torch.Tensor
) -> torch.Tensor:
    """
    A pair wise version of the cross entropy loss, see `sigmoid_cross_entropy_loss` for usage.

    Args:
        inputs (`torch.Tensor`):
            A tensor representing a mask.
        labels (`torch.Tensor`):
            A tensor with the same shape as inputs.
            Stores the binary classification labels for each element in inputs
            (0 for the negative class and 1 for the positive class).

    Returns:
        loss (`torch.Tensor`): The computed loss between each pairs.
    """

    height_and_width = inputs.shape[1]

    criterion = torch.nn.BCEWithLogitsLoss(reduction="none")
    # prevent overflow
    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs)) / height_and_width
    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs)) / height_and_width

    loss_pos = torch.matmul(cross_entropy_loss_pos, labels.T)
    loss_neg = torch.matmul(cross_entropy_loss_neg, (1 - labels).T)
    loss = loss_pos + loss_neg
    return loss
```

This helps prevent overflow at matmul computation with 16-bit precision or lower.

### Expected behavior

non-overflow ce-loss result

### PR DESCRIPTION
**Pull request by pjh4993**
# What does this PR do?

Fixes #25938 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
      #25938 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts

### CODE REVIEW

File: src/transformers/models/oneformer/modeling_oneformer.py [hunk 1]
```diff
@@ -193,13 +193,13 @@ def pair_wise_sigmoid_cross_entropy_loss(inputs: torch.Tensor, labels: torch.Ten
     height_and_width = inputs.shape[1]
 
     criterion = nn.BCEWithLogitsLoss(reduction="none")
-    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs))
-    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs))
+    # Prevent overflow when using torch 16bit autocast
```
Comment (amyeroberts) [inline src/transformers/models/oneformer/modeling_oneformer.py:196]: It might not be autocast - in some cases we explicitly set this   

```suggestion
    # Prevent overflow when using torch 16bit
```

File: src/transformers/models/oneformer/modeling_oneformer.py [hunk 2]
```diff
@@ -625,7 +625,10 @@ def sample_points_using_uncertainty(
 
         if num_random_points > 0:
             point_coordinates = torch.cat(
-                [point_coordinates, torch.rand(num_boxes, num_random_points, 2, device=logits.device)],
+                [
+                    point_coordinates,
+                    torch.rand(num_boxes, num_random_points, 2, device=logits.device),
+                ],
```
Comment (amyeroberts) [inline src/transformers/models/oneformer/modeling_oneformer.py:631]: Let's keep this on one line 

```suggestion
                [point_coordinates, torch.rand(num_boxes, num_random_points, 2, device=logits.device)],
```

File: src/transformers/models/mask2former/modeling_mask2former.py [hunk 3]
```diff
@@ -404,7 +404,11 @@ class Mask2FormerHungarianMatcher(nn.Module):
     """
 
     def __init__(
-        self, cost_class: float = 1.0, cost_mask: float = 1.0, cost_dice: float = 1.0, num_points: int = 12544
+        self,
+        cost_class: float = 1.0,
+        cost_mask: float = 1.0,
+        cost_dice: float = 1.0,
+        num_points: int = 12544,
```
Comment (amyeroberts) [inline src/transformers/models/mask2former/modeling_mask2former.py:411]: Let's keep the previous formatting and on one line 

```suggestion
        self, cost_class: float = 1.0, cost_mask: float = 1.0, cost_dice: float = 1.0, num_points: int = 12544
```

File: src/transformers/models/mask2former/modeling_mask2former.py [hunk 4]
```diff
@@ -584,7 +588,10 @@ def loss_labels(
             [target[j] for target, (_, j) in zip(class_labels, indices)]
         )  # shape of (batch_size, num_queries)
         target_classes = torch.full(
-            (batch_size, num_queries), fill_value=self.num_labels, dtype=torch.int64, device=pred_logits.device
+            (batch_size, num_queries),
+            fill_value=self.num_labels,
+            dtype=torch.int64,
+            device=pred_logits.device,
```
Comment (amyeroberts) [inline src/transformers/models/mask2former/modeling_mask2former.py:594]:
```suggestion
            (batch_size, num_queries), fill_value=self.num_labels, dtype=torch.int64, device=pred_logits.device
```

File: src/transformers/models/mask2former/modeling_mask2former.py [hunk 5]
```diff
@@ -734,7 +741,10 @@ def sample_points_using_uncertainty(
 
         if num_random_points > 0:
             point_coordinates = torch.cat(
-                [point_coordinates, torch.rand(num_boxes, num_random_points, 2, device=logits.device)],
+                [
+                    point_coordinates,
+                    torch.rand(num_boxes, num_random_points, 2, device=logits.device),
+                ],
```
Comment (amyeroberts) [inline src/transformers/models/mask2former/modeling_mask2former.py:747]:
```suggestion
                [point_coordinates, torch.rand(num_boxes, num_random_points, 2, device=logits.device)],
```

File: src/transformers/models/mask2former/modeling_mask2former.py [hunk 6]
```diff
@@ -1191,7 +1201,9 @@ def forward(
             all_hidden_states += (hidden_states.transpose(1, 0),)
 
         return BaseModelOutput(
-            last_hidden_state=hidden_states, hidden_states=all_hidden_states, attentions=all_attentions
+            last_hidden_state=hidden_states,
+            hidden_states=all_hidden_states,
+            attentions=all_attentions,
```
Comment (amyeroberts) [inline src/transformers/models/mask2former/modeling_mask2former.py:1206]:
```suggestion
            last_hidden_state=hidden_states, hidden_states=all_hidden_states, attentions=all_attentions
```

File: src/transformers/models/mask2former/modeling_mask2former.py [hunk 7]
```diff
@@ -1919,7 +1931,13 @@ def custom_forward(*inputs):
...
         hidden_states = hidden_states.transpose(1, 0)
         if not return_dict:
-            outputs = [hidden_states, all_hidden_states, attentions, intermediate, intermediate_mask_predictions]
+            outputs = [
+                hidden_states,
+                all_hidden_states,
+                attentions,
+                intermediate,
+                intermediate_mask_predictions,
+            ]
```
Comment (amyeroberts) [inline src/transformers/models/mask2former/modeling_mask2former.py:1940]:
```suggestion
            outputs = [hidden_states, all_hidden_states, attentions, intermediate, intermediate_mask_predictions]
```

File: src/transformers/models/mask2former/modeling_mask2former.py [hunk 8]
```diff
@@ -2008,7 +2026,12 @@ def __init__(self, hidden_size: int, num_heads: int, mask_feature_size: torch.Te
 
         self.mask_embedder = Mask2FormerMLPPredictionHead(self.hidden_size, self.hidden_size, mask_feature_size)
 
-    def forward(self, outputs: torch.Tensor, pixel_embeddings: torch.Tensor, attention_mask_target_size: int = None):
+    def forward(
+        self,
+        outputs: torch.Tensor,
+        pixel_embeddings: torch.Tensor,
+        attention_mask_target_size: int = None,
+    ):
```
Comment (amyeroberts) [inline src/transformers/models/mask2former/modeling_mask2former.py:2034]:
```suggestion
    def forward(self, outputs: torch.Tensor, pixel_embeddings: torch.Tensor, attention_mask_target_size: int = None):
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Just a comment on keeping the existing formatting as-is. Otherwise all looks good!
